src/cargo/ops/cargo_doc.rs

```diff
@@ -13,45 +13,53 @@ pub struct DocOptions<'a> {
 }
 
 pub fn doc(ws: &Workspace, options: &DocOptions) -> CargoResult<()> {
-    let package = ws.current()?;
+    let specs = options.compile_opts.spec.into_package_id_specs(ws)?;
+    let resolve = ops::resolve_ws_precisely(ws,
+                                            None,
+                                            options.compile_opts.features,
+                                            options.compile_opts.all_features,
+                                            options.compile_opts.no_default_features,
+                                            &specs)?;
+    let (packages, resolve_with_overrides) = resolve;
 
-    let spec = match options.compile_opts.spec {
-        ops::Packages::Packages(packages) => packages,
-        _ => {
-            // This should not happen, because the `doc` binary is hard-coded to pass
-            // the `Packages::Packages` variant.
-            bail!("`cargo doc` does not support the `--all` flag")
-        },
+    let mut pkgs = Vec::new();
+    if specs.len() > 0 {
+        for p in specs.iter() {
+            pkgs.push(packages.get(p.query(resolve_with_overrides.iter())?)?);
+        }
+    } else {
+        let root_package = ws.current()?;
+        pkgs.push(root_package);
     };
 
     let mut lib_names = HashSet::new();
     let mut bin_names = HashSet::new();
-    if spec.is_empty() {
-        for target in package.targets().iter().filter(|t| t.documented()) {
-            if target.is_lib() {
-                assert!(lib_names.insert(target.crate_name()));
-            } else {
-                assert!(bin_names.insert(target.crate_name()));
+    if specs.is_empty() {
```

Comment:
I think this may actually work best if we remove this condition. I don't quite recall why it exists, but I think it's best if we execute the below logic for all packages, not just the root package.